ocaml/xapi/nm.ml

```diff
@@ -520,32 +515,44 @@ let bring_pif_down ~__context ?(force=false) (pif: API.ref_PIF) =
             Net.Interface.set_persistent dbg bridge false;
             Db.PIF.set_currently_attached ~__context ~self:pif ~value:false;
 
-            (* If last tunnel with VxLAN protocol is unplugged, close VxLAN UDP port *)
-            match Xapi_pif_helpers.get_pif_type rc with
-            | Tunnel_access tunnel_ref ->
-              let protocol = Db.Tunnel.get_protocol ~__context ~self:tunnel_ref in
-              begin match protocol with
-              | `vxlan ->
-                let tunnels = Db.Tunnel.get_records_where ~__context ~expr: (
-                  Eq (Field "protocol", Literal (Record_util.tunnel_protocol_to_string `vxlan))
-                ) in
-                let rec vxlan_tunnel_still_plugged = function
-                | (_, tunnel_rec) :: tl ->
-                  let access_pif_ref = tunnel_rec.API.tunnel_access_PIF in
-                  let plugged = Db.PIF.get_currently_attached ~__context ~self:access_pif_ref in
-                  if plugged then
-                    true
-                  else
-                    vxlan_tunnel_still_plugged tl
-                | [] -> false
-                in
-                let no_more_vxlan = not (vxlan_tunnel_still_plugged tunnels) in
-                if no_more_vxlan then begin
-                  debug "Last VxLAN tunnel was closed, closing VxLAN UDP port";
-                  ignore @@ Helpers.call_script !Xapi_globs.firewall_port_config_script ["close"; "4789"; "udp"]
-                end;
-              | `gre -> ()
-              end;
-            | _ -> ()
-          )
-    )
+            cleanup_f ()
+        )
+      in
+      match get_pif_topo ~__context ~pif_rec:rc with
+      | Network_sriov_logical _ :: _ ->
+        Xapi_network_sriov_helpers.sriov_bring_down ~__context ~self:pif
+      | VLAN_untagged _ :: Network_sriov_logical _ :: _ ->
+        Db.PIF.set_currently_attached ~__context ~self:pif ~value:false
+      | Tunnel_access tunnel_ref :: _ ->
+          (* If last tunnel with VxLAN protocol is unplugged, close VxLAN UDP port *)
+          let maybe_close_port () =
+            let host = Db.PIF.get_host ~__context ~self:pif in
```

Comment:
You have the record, so avoid the DB call:

     let host = rc.API.pIF_host in